cpp/src/arrow/parquet/writer.cc

```diff
@@ -143,6 +153,82 @@ Status FileWriter::Close() {
 
 FileWriter::~FileWriter() {}
 
+// Create a slice of a PrimitiveArray.
+//
+// This method is specially crafted for WriteFlatTable and assumes the following:
+//  * chunk_size is a multiple of 512
```

Comment:
It might pay to just give a little bit more detail (size of 512 to make the produced array conform to alignment/padding issues).  We should file a JIRA to make sure DCHECK is more widely available.

Until we have more clarity on this, I like enforcing the assumption (it is generally less painful to remove these assertions then add them in after the fact).    I've added some ramblings on possible solutions below.  We should open a JIRA to hash this out but here is one possible proposal.  But I've added notes below.

One option is to expose IsComforming/MakeConforming methods on the arrays and let algorithms decide based the relative expense of memcpy vs SIMD efficiency, whether to pay the expense of doing the conversion.   The one thing that can get tricky here is are there different levels of conforming (e.g. padded but not aligned; non-padded, not aligned but a bitmask still has elements beyond the size of array marked as invalid, etc).  For simplicity I would say there is likely just two states and call it good.

It would be required for IPC that the Arrays are conforming.